File: src/convertData.js
class Edge {
    from;
    to;
    cost;
  
    constructor(from, to, cost) {
      this.from = from;
      this.to = to;
      this.cost = cost;
    }
  
  }
  
  
  class UnionFind {
    parent = [];
  
    constructor(n) {
      for(let i = 0; i < n; i++) {
        this.parent.push(i);
      }
    }
  
    root(x) {
      if(this.parent[x] === x) return x;
      this.parent[x] = this.root(this.parent[x]);
      return this.parent[x];
    }
  
    unite(x, y) {
      let rx = this.root(x);
      let ry = this.root(y);
      if(rx === ry) return 0;
      this.parent[rx] = ry;
    }
  
    same(x, y) {
      let rx = this.root(x);
      let ry = this.root(y);
  
      if(rx === ry) {
        return true;
      } else {
        return false;
      }
    }
  
}

const convertData = () => {
    const res = [];
    const response1 = await fetch('data/jaccard_data_100.json');
    const response2 = await fetch('data/difficulty_data_100.json');

    const jaccard_data = await response1.json();
    const diff_data = await response2.json();
    
    const len = jaccard_data.length;

    const es = [];
    const uf = new UnionFind(len);
    for(let i = 0; i < len; i++) {
        const obj = {};
        obj['NodeName'] = diff_data[i]['tag'];
        obj['diff'] = diff_data[i]['difficulty'];
        obj['ID'] = i+1;
        obj['url'] = `https://qiita.com/tags/${diff_data[i]['tag'].toUpperCase()}`;
        obj['childNode'] = [];
        res.push(obj);
    }

    for(let i = 0; i < len; i++) {
        for(let j = 0; j < len; j++) {
            if(i != j) {
                es.push(new Edge(i, j, jaccard_data[i]['jaccard'][j]));
            }
        }
    }

    es.sort((a, b) => {
        return b.cost - a.cost;
    })

    for(let i = 0; i < 100; i++) {
        console.log(es[i]);
    }


    for(const item of es) {
        if(uf.same(item.to, item.from) === false) {
            uf.unite(item.to, item.from);
            const to, from;
            if(diff_data[item.to]['diffculty'] >= diff_data[item.from]['diffculty']) {
                to = item.to;
                from = item.from;
            } else {
                to = item.from;
                from = item.to;
            }

            res[from]['childNode'].push(to+1);

        }
    }

    for(let i = 0; i < 100; i++) {
        console.log(res[i]);
    }
    return res;
}

File: src/convertData.py
import json
    
class UnionFind :
    parent = []
  
    def __init__(self, n): 
      for i in range(n): 
        self.parent.append(i)
    
  
    def root(self, x):
      if(self.parent[x] == x):
           return x
      self.parent[x] = self.root(self.parent[x])
      return self.parent[x]
    
  
    def unite(self, x, y): 
      rx = self.root(x)
      ry = self.root(y)
      if(rx == ry):
          return 0
      self.parent[rx] = ry
    
  
    def same(self, x, y):
      rx = self.root(x)
      ry = self.root(y)
  
      if(rx == ry):
        return True
      else:
        return False

def main():
    res = []
    f1 = open('jaccard_data_100.json' ,'r', encoding="utf-8_sig")
    f2 = open('difficulty_data_100.json','r', encoding="utf-8_sig")
    jaccard_data = json.load(f1)
    diff_data = json.load(f2)
    f1.close()
    f2.close()
    
    leng = len(jaccard_data)

    es = []
    uf = UnionFind(leng)

    for i in range(leng):
        obj = {}
        obj['nodeName'] = diff_data[i]['tag']
        obj['diff'] = diff_data[i]['difficulty']
        obj['ID'] = i+1
        obj['url'] = 'https://qiita.com/tags/{}'.format(obj['nodeName'].lower())
        obj['childNode'] = []
        res.append(obj)
    
    for i in range(leng):
        for j in range(leng):
            if(i != j): 
                obj = {}
                obj['From'] = i
                obj['To'] = j
                obj['cost'] = jaccard_data[i]['jaccard'][j]
                es.append(obj)
               
        
    es = sorted(es, key = lambda x:x['cost'], reverse=True)
    
    #for i in range(20):
    #    print(es[i])

    for item in es:
      if(uf.same(item['To'], item['From']) == False):
            uf.unite(item['To'], item['From'])
            To = -1 
            From = -1
            if(diff_data[item['To']]['difficulty'] >= diff_data[item['From']]['difficulty']):
                To = item['To']
                From = item['From']
            else:
                To = item['From']
                From = item['To']
            res[From]['childNode'].append(To+1)

    for i in range(20):
      print(res[i])
    
    with open('new_data.json', 'w', encoding='utf-8') as f:
      f.write(json.dumps(res, indent=2))

if __name__ == '__main__':
    main()